### PR TITLE
REFACTOR-#7294: Reduce access of methods _modin_frame methods from query compiler objects

### LINKED ISSUE #7294: REFACTOR: Minimize the access of methods _modin_frame methods from ._query_compiler layer
Minimize the access of methods _modin_frame methods from ._query_compiler layer and adding these functions to query_compiler API so they could be called directly without directly exposing _modin_frame.

This changes are required for #7259 and creating a separate issue and PR as suggested  by @anmyachev in https://github.com/modin-project/modin/pull/7259#discussion_r1616020046_

### PR DESCRIPTION
**Pull request by arunjose696**
## What do these changes do?

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [ ] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [ ] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [ ] signed commit with `git commit -s` 
- [ ] Resolves #7294  
- [ ] tests added and passing
- [ ] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/core/storage_formats/base/query_compiler.py [hunk 1]
```diff
@@ -4521,6 +4521,60 @@ def has_multiindex(self, axis=0):
...
 
+    @property
+    def has_materialized_dtypes(self) -> bool:
+        """
+        Check if the undelying dataframe has materialized dtypes.
+
+        Returns
+        -------
+        bool
+            True if if the undelying dataframe and False otherwise.
```
Comment (anmyachev) [inline modin/core/storage_formats/base/query_compiler.py:4532]:
```suggestion
```

File: modin/core/storage_formats/base/query_compiler.py [hunk 2]
```diff
@@ -4521,6 +4521,60 @@ def has_multiindex(self, axis=0):
         assert axis == 1
         return isinstance(self.columns, pandas.MultiIndex)
 
+    @property
+    def has_materialized_dtypes(self) -> bool:
```
Comment (anmyachev) [inline modin/core/storage_formats/base/query_compiler.py:4525]: By analogy
```suggestion
    def frame_has_materialized_dtypes(self) -> bool:
```

**Review by anmyachev — CHANGES_REQUESTED** [review]

File: modin/core/storage_formats/base/query_compiler.py (hunk 2 above)
Comment (arunjose696) [reply modin/core/storage_formats/base/query_compiler.py]: Done

File: modin/core/storage_formats/base/query_compiler.py [hunk 3]
```diff
@@ -4521,6 +4521,59 @@ def has_multiindex(self, axis=0):
         assert axis == 1
         return isinstance(self.columns, pandas.MultiIndex)
 
+    @property
+    def frame_has_materialized_dtypes(self) -> bool:
```
Comment (YarShev) [inline modin/core/storage_formats/base/query_compiler.py:4525]: @anmyachev, @arunjose696, why didn't we add similar methods for operating on columns?

Comment (anmyachev) [reply modin/core/storage_formats/base/query_compiler.py:4525]: Apparently the relevant code was simply not touched when working on the small query compiler. There is no reason not to do this for columns.

Comment (YarShev) [reply modin/core/storage_formats/base/query_compiler.py:4525]: I would propose to do this for consistency. @arunjose696, would you take on this?

**Pull request merged**